pkg/virt-controller/watch/job_test.go

```diff
@@ -51,55 +63,133 @@ var _ = Describe("Migration", func() {
 
-		stopChan = make(chan struct{})
-		lw = framework.NewFakeControllerSource()
 		jobCache = cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc, nil)
-
-		_, jobController = NewJobControllerWithListWatch(vmService, nil, lw, restClient)
+		jobQueue = workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
 
 		vm = kvirtv1.NewMinimalVM("test-vm")
 		vm.Status.Phase = kvirtv1.Migrating
 		vm.GetObjectMeta().SetLabels(map[string]string{"a": "b"})
 
-		// Start the controller
-		jobController.StartInformer(stopChan)
-		go jobController.Run(1, stopChan)
+		dispatch = NewJobControllerDispatch(vmService, restClient)
+		migration = kvirtv1.NewMinimalMigration("test-migration", "test-vm")
+		job = &corev1.Pod{
+			ObjectMeta: corev1.ObjectMeta{
+				Labels: map[string]string{
+					kvirtv1.AppLabel:       "migration",
+					kvirtv1.DomainLabel:    "test-vm",
+					kvirtv1.MigrationLabel: migration.ObjectMeta.Name,
+				},
+			},
+			Status: corev1.PodStatus{
+				Phase: corev1.PodSucceeded,
+			},
+		}
+
 	})
 
-	Context("Running job with migration labels and one success", func() {
-		It("should update the VM to Running", func(done Done) {
+	Context("Running job", func() {
+		It("Success should update the VM to Running", func(done Done) {
+			// Register the expected REST call
+			server.AppendHandlers(
+				handlerToFetchTestVM(vm),
+				handlerToUpdateTestVM(vm),
+				handlerToFetchTestMigration(migration),
+				handlerToUpdateTestMigration(migration, kvirtv1.MigrationSucceeded),
+			)
+
+			doExecute()
+			Expect(len(server.ReceivedRequests())).To(Equal(4))
+			close(done)
+		}, 10)
 
-			migration := kvirtv1.NewMinimalMigration("test-migration", "test-vm")
-			job := &corev1.Pod{
-				ObjectMeta: corev1.ObjectMeta{
-					Labels: map[string]string{
-						kvirtv1.DomainLabel:    "test-vm",
-						kvirtv1.MigrationLabel: migration.ObjectMeta.Name,
-					},
-				},
-				Status: corev1.PodStatus{
-					Phase: corev1.PodSucceeded,
-				},
-			}
+		It("should have valid listSelector", func() {
+
+			f := fields.Set{"Status.Phase": string(corev1.PodSucceeded)}
+			matches := listOptions.FieldSelector.Matches(f)
+			Expect(matches).Should(BeTrue())
+
+			s := labels.Set(job.Labels)
+			matches = listOptions.LabelSelector.Matches(s)
+			Expect(matches).Should(BeTrue())
+		}, 100)
 
+		It("Error calling Fetch VM should requeue", func(done Done) {
+			// Register the expected REST call
+			server.AppendHandlers(
+				handlerToFetchTestVMAuthError(vm),
+			)
+
+			doExecute()
+
+			Expect(len(server.ReceivedRequests())).To(Equal(1))
+			Eventually(jobQueue.Len()).Should(Equal(1))
+			close(done)
+		}, 10)
+
+		It("Error Updating VM should requeue", func(done Done) {
+			// Register the expected REST call
+			server.AppendHandlers(
+				handlerToFetchTestVM(vm),
+				handlerToUpdateTestVMAuthError(vm),
+			)
+
+			doExecute()
+
+			Expect(len(server.ReceivedRequests())).To(Equal(2))
+			Eventually(jobQueue.Len()).Should(Equal(1))
```

Comment:
A check if we have a history of the key is missing: `queue.NumRequeues(key) == 1`